DCOS CLI Integration

huntc opened the pull request:
This commit provides the changes required in order for the "conduct" CLI to be integrated with the DC/OS CLI such that you can "dcos conduct ..." and the ConductR commands are channelled via DC/OS's admin gateway.

huntc committed:
DCOS CLI Integration

This commit provides the changes required in order for the "conduct" CLI to be integrated with the DC/OS CLI such that you can "dcos conduct ..." and the ConductR commands are channelled via DC/OS's admin gateway.

huntc commented:
Observational testing:

```
ChristohersMBP2:conductr-cli huntc$ dcos 
Command line utility for the Mesosphere Datacenter Operating
System (DC/OS). The Mesosphere DC/OS is a distributed operating
system built around Apache Mesos. This utility provides tools
for easy management of a DC/OS installation.

Available DC/OS commands:

    auth            Authenticate to DC/OS cluster
    conduct         Lightbend ConductR sub commands. Type 'dcos conduct' to see more.
    config          Manage the DC/OS configuration file
    help            Display help information about DC/OS
    marathon        Deploy and manage applications to DC/OS
    node            Administer and manage DC/OS cluster nodes
    package         Install and manage DC/OS software packages
    service         Manage DC/OS services
    task            Manage DC/OS tasks

Get detailed command description with 'dcos <command> --help'.
ChristohersMBP2:conductr-cli huntc$ 
ChristohersMBP2:conductr-cli huntc$ 
ChristohersMBP2:conductr-cli huntc$ dcos conduct info
ID       NAME        #REP  #STR  #RUN
c14bd62  visualizer     1     0     0
```

huntc commented:
Failing because of the DC/OS Python library unable to be found.

viktorklang commented on conductr_cli/conduct_main.py:
```diff
@@ -5,15 +5,22 @@
-def add_ip_and_port(sub_parser):
+def add_scheme_ip_port_and_base_path(sub_parser):
+    sub_parser.add_argument('--scheme',
+                            help='The optional ConductR scheme, defaults to `http`',
```
Shouldn't something as important as the orchestrator API default to https? (perhaps that's for the future?)

huntc replied:
Agreed. TLS support for the control protocol is planned. Meanwhile our DC/OS integration here will provide TLS support.

markusjura commented on conductr_cli/conduct_main.py:
```diff
@@ -270,19 +299,47 @@ def get_custom_settings(args):
+    parser = build_parser(dcos_mode)
     argcomplete.autocomplete(parser)
-    args = parser.parse_args(_args) if _args else parser.parse_args()
+    args = parser.parse_args(_args[1:])
```
We are removing here always the first argument. Also in non DCOS mode. Does that makes sense?

huntc replied:
Yes. This is what the previous code was actually doing under the hood.

markusjura replied:
👍

markusjura commented on conductr_cli/conduct_main.py:
```diff
@@ -270,19 +299,47 @@ def get_custom_settings(args):
+    args = parser.parse_args(_args[1:])
     if not vars(args).get('func'):
-        parser.print_help()
+        if not vars(args).get('dcos_info'):
```
For readability reasons I'd would swap the if and else clause:

```
        if vars(args).get('dcos_info'):
            print('Lightbend ConductR sub commands. Type \'dcos conduct\' to see more.')
            exit(0)            
        else:
            parser.print_help()
```

huntc replied:
Why would that be more readable?

markusjura replied:
We are saving a "not". In a if / else statement you should never start with the negative, but always with the positive case.

huntc replied:
There's also an argument to show the happy path, but I'll go with it.

markusjura commented on conductr_cli/sandbox_run.py:
```diff
@@ -203,7 +204,7 @@ def wait_for_conductr(args, current_retry, max_retries, interval):
         url = conduct_url.url('members', conduct_args)
         try:
-            requests.get(url, timeout=DEFAULT_HTTP_TIMEOUT, headers=conduct_url.request_headers(conduct_args))
+            conduct_request.get(False, ConductArgs.ip, url, timeout=DEFAULT_HTTP_TIMEOUT)
```
Always specify the named argument when passing in a boolean:

```
conduct_request.get(dcos_mode=False, ConductArgs.ip, url, timeout=DEFAULT_HTTP_TIMEOUT)
```

huntc replied:
ok, but I do then have to specify the other arg names.

markusjura replied:
Your decision. Not critical.

markusjura commented on conductr_cli/test/conduct_load_test_base.py:
```diff
@@ -64,23 +63,20 @@ def base_test_success(self):
-        request_headers_mock.assert_called_with(input_args)
         http_method.assert_called_with(self.default_url, files=self.default_files, timeout=LOAD_HTTP_TIMEOUT,
-                                       headers=self.mock_headers)
+                                       headers={'Host': '127.0.0.1'})
```
Why is this change necessary?

huntc replied:
We were previously checking the host header being set up via another mock. The other mock was no longer available due to its function having been replaced. The code should also be much simpler now.

markusjura replied:
👍

markusjura commented:
> conduct init needs to create a soft link to the `conduct.py` file from `~/.dcos/subcommands/conductr/env/bin/dcos-conduct`

What do you mean by that? We only have a `sandbox init` command but this should not be involved in DCOS mode.

huntc commented:
> What do you mean by that? We only have a sandbox init command but this should not be involved in DCOS mode.

In order to make the DC/OS CLI aware of the ConductR CLI, you have to tell it. Ordinarily, this is done via `dcos install --cli <package>`, but we don't have a universal package available (yet). So, meanwhile then, `conduct init` will automatically set things up.

I just realised though that we have no `conduct init`, so we'll go with your IM suggestion of a new `conduct setup-dcos`.

huntc commented:
Waiting on a new DCOS lib release. PR tests will fail here until the following PR is merged and there's a new release of the DCOS lib: https://github.com/dcos/dcos-cli/pull/823

huntc committed:
Pass instance info to the sandbox

We can then perform activities within the sandbox in relation to the instance number, and the number of target instances.

huntc committed:
CLI test commands

As a result of the IP never being able to be None.

huntc committed:
Adds DC/OS lib

Eliminates our dependency on the DC/OS lib until the day that https://github.com/dcos/dcos-cli/pull/823 gets merged and we have a new lib.

markusjura commented:
Manual integration test with `http://dcos-master.lightbend.com` was successful.
